### PR TITLE
[AIRFLOW-6586] Improvements to gcs sensor

### PR DESCRIPTION
**Pull request by jaketf**
refactors GoogleCloudStorageUploadSessionCompleteSensor to use set instead of number of objects

add poke mode only decorator: this is important because stateful sensors will break in reschedule mode.

assert that poke_mode_only applied to child of BaseSensorOperator

---
Issue link: [AIRFLOW-6586](https://issues.apache.org/jira/browse/AIRFLOW-6586)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/sensors/decorators/poke_mode_only.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    that instances of this class are only safe to use poke mode.
+    
+    Will decorate all methods in the class to assert they did not change
+    the mode from 'poke'.
+    
+    :param cls: BaseSensor class to enforce methods only use 'poke' mode.
+    :type cls: type
+    """
+    def decorate(cls):
+        assert issubclass(cls, BaseSensorOperator)
```
Comment (turbaszek) [inline airflow/sensors/decorators/poke_mode_only.py]: Please use explicit exception. We've agreed to not use asserts in production code:
https://lists.apache.org/thread.html/bcf2d23fcd79e21b3aac9f32914e1bf656e05ffbcb8aa282af497a2d%40%3Cdev.airflow.apache.org%3E

**Review by turbaszek — CHANGES_REQUESTED** [review]
I like the idea! That's something we discussed with @mik-laj last week. Just a comment about our lovely `assert` :D

**Review by turbaszek — APPROVED** [review]

**Comment by turbaszek** [comment]
@jaketf any update?

**Comment by jaketf** [comment]
Apologies for letting this go stale I will revive this later this week.

**Comment by jaketf** [comment]
@turbaszek 
The static CI check seems to be unrelated about docker linting.
The build docs CI check is complaining that I don't document this decorator as a sensor. 
How do you suggest we remedy this? is there a more appropriate directory for this to fall under in the code base?

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 2]
```diff
@@ -189,12 +190,14 @@ def get_time():
...
 
+@poke_mode_only
 class GCSUploadSessionCompleteSensor(BaseSensorOperator):
     """
     Checks for changes in the number of objects at prefix in Google Cloud Storage
     bucket and returns True if the inactivity period has passed with no
-    increase in the number of objects. Note, it is recommended to use reschedule
-    mode if you expect this sensor to run for hours.
+    increase in the number of objects. Note, this sensor will no behave correctly
+    in reschedule mode, as the state of the listed objects in the GCS bucket will
```
Comment (RosterIn) [inline airflow/providers/google/cloud/sensors/gcs.py:198]:
```suggestion
    increase in the number of objects. Note, this sensor will not behave correctly
    with reschedule mode, as the state of the listed objects in the GCS bucket will
```

**Comment by turbaszek** [comment]
@jaketf please rebase, it seems that we had some problems, from devlist:

> Hello everyone, I just merged a change with failing DockerLint (due to
yesterday's released hadolint new version). Please rebase to the
latest master. The fix pins to the version released yesterday so it
should not happen again.

https://github.com/apache/airflow/pull/8485

**Comment by jaketf** [comment]
@turbaszek this is the first decorator in this dir. How do you think we should document this? 
In the same place the docs ci check is complaining about ? operators-and-hooks-ref.rst? 
Should I create a new docs file for sensor decorators?

File: airflow/sensors/decorators/poke_mode_only.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    :param cls: BaseSensor class to enforce methods only use 'poke' mode.
+    :type cls: type
+    """
+    def decorate(cls):
+        if not issubclass(cls, BaseSensorOperator):
+            raise ValueError("poke_mode_only decorator should only be "
+                             + "applied to subclasses of BaseSensorOperator,"
+                             + f" got:{cls}.")
+        for method in inspect.getmembers(cls, inspect.isfunction):
```
Comment (mik-laj) [inline airflow/sensors/decorators/poke_mode_only.py]: Is it really necessary to decorate all methods? Is it not enough to create the read-only `poke` attribute?

Something similar to code below:
```python
    @property
    def mode(self):
        return "poke"

    @mode.setter
    def mode(self, value):
        if value != "poke":
            raise ValueError()

```

Comment (mik-laj) [reply airflow/sensors/decorators/poke_mode_only.py]: or as a decorator
```python
def poke_only(cls_type):
    def mode_getter(_):
        return "poke"
    
    cls_type.mode = property(mode_getter)

    def mode_setter(_, value):
        if value != "value":
            raise ValueError()
        
    cls_type.mode.setter(mode_setter)
```
Both examples are not tested but should work.

**Comment by mik-laj** [comment]
This decorator should be in the same module as BaseSensorOperator

If you want to write some documentation, you can add information about it to the guide on writing operators.
https://airflow.readthedocs.io/en/latest/howto/custom-operator.html

**Comment by jaketf** [comment]
@mik-laj thanks it does make more sense in that module. 
looks like those docs don't really touch on sensors so `poke_mode_only` docs there would've seem out of place.

I've added a short blurb on sensors to that custom operators page.

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 4]
```diff
@@ -243,45 +246,57 @@ def __init__(self,
 
         self.bucket = bucket
         self.prefix = prefix
+        if inactivity_period < 0:
+            raise ValueError(
+                f"inactivity_period must be non-negative")
```
Comment (turbaszek) [inline airflow/providers/google/cloud/sensors/gcs.py]:
```suggestion
            raise ValueError("inactivity_period must be non-negative")
```

File: airflow/sensors/base_sensor_operator.py [hunk 5]
```diff
@@ -176,3 +179,60 @@ def deps(self):
...
+    the mode from 'poke'.
+
+    :param cls: BaseSensor class to enforce methods only use 'poke' mode.
+    :type cls: type
+    """
+    def decorate(cls):
+        if not issubclass(cls, BaseSensorOperator):
+            raise ValueError("poke_mode_only decorator should only be "
+                             + "applied to subclasses of BaseSensorOperator,"
+                             + f" got:{cls}.")
```
Comment (turbaszek) [inline airflow/sensors/base_sensor_operator.py]:
```suggestion
            raise ValueError(f"poke_mode_only decorator should only be "
                                         f"applied to subclasses of BaseSensorOperator,"
                                         f" got:{cls}.")
```

Comment (turbaszek) [reply airflow/sensors/base_sensor_operator.py]: The `+` isn't necessary

File: airflow/sensors/base_sensor_operator.py [hunk 6]
```diff
@@ -176,3 +179,60 @@ def deps(self):
         if self.reschedule:
             return BaseOperator.deps.fget(self) | {ReadyToRescheduleDep()}
         return BaseOperator.deps.fget(self)
+
+
+def poke_mode_only(cls):
```
Comment (mik-laj) [inline airflow/sensors/base_sensor_operator.py:182]: Why do we need such a complex decorator?
https://github.com/apache/airflow/pull/7197#discussion_r419218559

**Comment by jaketf** [comment]
Build docs is sad, but I do not understand the error message.
Is there a way I can catch things like this sooner than CI checks? my pre-commit was clean.
```
/opt/airflow/docs/_api/airflow/sensors/base_sensor_operator/index.rst:108: WARNING: Error in "function" directive:
invalid option data: extension option field name may not contain multiple words.

.. function:: _poke_mode_only_func_decorator(method)
   Decorator that raises an error if method changed mode to anything but
   'poke'.function
   :param method: BaseSensor class to enforce methods only use 'poke' mode.
   :type method: function
```

File: airflow/sensors/decorators/poke_mode_only.py (hunk 3 above)
Comment (jaketf) [reply airflow/sensors/decorators/poke_mode_only.py]: I agree decorating all methods is overly complicated.

The first requires that every user implementing a poke mode only operator do that in their implementation. 

I'll give the second one a shot and see if it passes my existing tests.

File: airflow/sensors/base_sensor_operator.py [hunk 7]
```diff
@@ -176,3 +179,60 @@ def deps(self):
...
+        return cls
+
+    return decorate(cls)
+
+
+def _poke_mode_only_func_decorator(method):
+    """
+    Decorator that raises an error if method changed mode to anything but
+    'poke'.function
+    :param method: BaseSensor class to enforce methods only use 'poke' mode.
```
Comment (mik-laj) [inline airflow/sensors/base_sensor_operator.py]:
```suggestion

    :param method: BaseSensor class to enforce methods only use 'poke' mode.
```

**Comment by mik-laj** [comment]
@jaketf Unfortunately, you need to build documentation to detect such problems. We don't have pre-commit yet which check the documentation.
This issue looks related: https://github.com/apache/airflow/issues/8091

File: airflow/sensors/decorators/poke_mode_only.py (hunk 3 above)
Comment (jaketf) [reply airflow/sensors/decorators/poke_mode_only.py]: Yeah that passes tests nicely.
Thanks so much for this suggestion it's so much cleaner.

**Comment by jaketf** [comment]
Please re-run travis.
It appears the failure is unrelated in TestKubernetesExecutor
https://travis-ci.org/github/apache/airflow/jobs/683599411#L2002

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 8]
```diff
@@ -226,14 +228,15 @@ class GCSUploadSessionCompleteSensor(BaseSensorOperator):
 
     template_fields = ('bucket', 'prefix')
     ui_color = '#f0eee4'
+    previous_objects: Set[str]
```
Comment (kaxil) [inline airflow/providers/google/cloud/sensors/gcs.py]: Why is this here?

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 9]
```diff
@@ -226,14 +228,15 @@ class GCSUploadSessionCompleteSensor(BaseSensorOperator):
...
 
     @apply_defaults
     def __init__(self,
                  bucket: str,
                  prefix: str,
                  inactivity_period: float = 60 * 60,
                  min_objects: int = 1,
-                 previous_num_objects: int = 0,
+                 previous_objects: Optional[Set[str]] = None,
```
Comment (kaxil) [inline airflow/providers/google/cloud/sensors/gcs.py:238]: This is a change in the signature that should need a note in Updating.md

File: docs/howto/custom-operator.rst [hunk 10]
```diff
@@ -206,3 +206,27 @@ Define an operator extra link
...
+
+Reschedule mode comes with a caveat that your sensor cannot maintain internal state
+between rescheduled executions. In this case you should decorate your sensor with
+:meth:`airflow.sensors.base_sensor_operator.poke_mode_only`. This will let users know
+that your sensor is not suitable for use with reschedule mode.
+
+An example of a sensor that keeps internal state and cannot be used with reschedule mode
+is :class:`airflow.providers.google.cloud.sensors.gcs.GCSUploadSessionCompleteSensor`.
+It polls the number of objects at a prefix (this number is the internal state of the sensor)
+and succeeds when there a certain amount of time has passed without the number of objects changing.
```
Comment (kaxil) [inline docs/howto/custom-operator.rst:232]: Nice, thanks for the docs

File: airflow/providers/google/cloud/sensors/gcs.py (hunk 9 above)
Comment (jaketf) [reply airflow/providers/google/cloud/sensors/gcs.py:238]: done.

File: airflow/providers/google/cloud/sensors/gcs.py (hunk 8 above)
Comment (kaxil) [reply airflow/providers/google/cloud/sensors/gcs.py]: Is it used somewhere ?

Comment (jaketf) [reply airflow/providers/google/cloud/sensors/gcs.py]: good catch, no good reason.
removed.

**Review by kaxil — APPROVED** [review]

**Review by turbaszek — APPROVED** [review]

**Comment by jaketf** [comment]
@turbaszek seems like only quarantined test failing. Any blocker to merge?

File: UPDATING.md [hunk 11]
```diff
@@ -118,6 +118,29 @@ plugins =
...
+- Old Behavior: This constructor used to optionally take ``previous_num_objects: int``.
+- New replacement constructor kwarg: ``previous_objects: Optional[Set[str]]``.
+
+Most users would not specify this argument because the bucket begins empty
+and the user wants to treat any files as new.
+
+Example of Updating usage of this sensor:
+Users who used to call:
+
+``GCSUploadSessionCompleteSensor('my_bucket', 'my_prefix', previous_num_objects=1)``
```
Comment (mik-laj) [inline UPDATING.md]: Operators must use keywords arguments. See: @apply_defaults.

**Review by mik-laj — APPROVED** [review]
